lxd/instances_put.go

```diff
@@ -0,0 +1,319 @@
+	switch action {
+	case shared.Start:
+		opType = db.OperationInstanceStart
+		do = func(op *operations.Operation) error {
+			failures := map[string]error{}
+			failuresLock := sync.Mutex{}
+			wgAction := sync.WaitGroup{}
+
+			for _, inst := range instances {
+				wgAction.Add(1)
+				go func(inst instance.Instance) {
+					defer wgAction.Done()
+
+					err = inst.Start(raw.Stateful)
+					if err != nil {
+						failuresLock.Lock()
+						failures[inst.Name()] = err
+						failuresLock.Unlock()
+					}
+				}(inst)
+			}
+
+			wgAction.Wait()
+			return coalesceErrors(failures)
+		}
+	case shared.Stop:
+		opType = db.OperationInstanceStop
+		if raw.Stateful {
+			do = func(op *operations.Operation) error {
+				failures := map[string]error{}
+				failuresLock := sync.Mutex{}
+				wgAction := sync.WaitGroup{}
+
+				for _, inst := range instances {
+					wgAction.Add(1)
+					go func(inst instance.Instance) {
+						defer wgAction.Done()
+
+						err = inst.Stop(raw.Stateful)
+						if err != nil {
+							failuresLock.Lock()
+							failures[inst.Name()] = err
+							failuresLock.Unlock()
+						}
+					}(inst)
+				}
+
+				wgAction.Wait()
+				return coalesceErrors(failures)
+			}
+		} else if raw.Timeout == 0 || raw.Force {
+			do = func(op *operations.Operation) error {
+				failures := map[string]error{}
+				failuresLock := sync.Mutex{}
+				wgAction := sync.WaitGroup{}
+
+				for _, inst := range instances {
+					wgAction.Add(1)
+					go func(inst instance.Instance) {
+						defer wgAction.Done()
+
+						err = inst.Stop(false)
+						if err != nil {
+							failuresLock.Lock()
+							failures[inst.Name()] = err
+							failuresLock.Unlock()
+						}
+					}(inst)
+				}
+
+				wgAction.Wait()
+				return coalesceErrors(failures)
+			}
+		} else {
+			do = func(op *operations.Operation) error {
+				failures := map[string]error{}
+				failuresLock := sync.Mutex{}
+				wgAction := sync.WaitGroup{}
+
+				for _, inst := range instances {
+					wgAction.Add(1)
+					go func(inst instance.Instance) {
+						defer wgAction.Done()
+
+						if inst.IsFrozen() {
+							err := inst.Unfreeze()
+							if err != nil {
+								failuresLock.Lock()
+								failures[inst.Name()] = err
+								failuresLock.Unlock()
+								return
+							}
+						}
+
+						err = inst.Shutdown(time.Duration(raw.Timeout) * time.Second)
+						if err != nil {
+							failuresLock.Lock()
+							failures[inst.Name()] = err
+							failuresLock.Unlock()
+						}
+					}(inst)
+				}
+
+				wgAction.Wait()
+				return coalesceErrors(failures)
+			}
+		}
+	case shared.Restart:
+		opType = db.OperationInstanceRestart
+		do = func(op *operations.Operation) error {
+			failures := map[string]error{}
+			failuresLock := sync.Mutex{}
+			wgAction := sync.WaitGroup{}
+
+			for _, inst := range instances {
+				wgAction.Add(1)
+				go func(inst instance.Instance) {
+					defer wgAction.Done()
+
+					ephemeral := inst.IsEphemeral()
```

Comment:
@stgraber not strictly related to the PR, but it seems like this logic should be moved into the Instance `Restart()` functions (or a common function called from the per-driver Restart() functions), as this feels like internal logic that the instance should be taking care of, rather than the 'user' of the Restart() function having to know about (I write this as someone who wouldn't have thought to handle the ephemeral case for Restart). What do you think?